File: src/main/java/me/rsman/BetterMinecraftCore/configs/models/BmcShapedCraft.java
package me.rsman.BetterMinecraftCore.configs.models;

import me.rsman.BetterMinecraftCore.BetterMinecraftCore;
import me.rsman.BetterMinecraftCore.Managers.ConfigManager;
import me.rsman.BetterMinecraftCore.configs.containers.BmcCraftContainer;
import org.bukkit.Bukkit;
import org.bukkit.NamespacedKey;
import org.bukkit.inventory.RecipeChoice;
import org.bukkit.inventory.ShapedRecipe;

import java.util.*;

public class BmcShapedCraft {
    private List<String> shape;
    private String result;
    private String name;
    private String key;

    public BmcShapedCraft() {}

    public List<String> getShape() {
        return shape;
    }

    public void setShape(List<String> shape) {
        this.shape = shape;
    }

    public String getResult() {
        return result;
    }

    public void setResult(String result) {
        this.result = result;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getKey() {
        return key;
    }

    public void setKey(String key) {
        this.key = key;
    }

    public void registerCraft(){
        String[] shapeArr = {"","",""};
        Map<Character, RecipeChoice> ingredients = new HashMap<>();

        assert result != null;
        Map.Entry<Character, RecipeChoice> resultItemFromScheme = BmcCraftSubContainer.convertSchemeToRecipeChoice(result, 'X');
        if(resultItemFromScheme == null || resultItemFromScheme.getValue() == null){
            BetterMinecraftCore.getInstance().getLogger().warning("§cskipping shaped recipe §3"+name+" "+key+"§c, result is invalid or not registered");
            return;
        }
        ShapedRecipe rec;
        if(resultItemFromScheme.getValue() instanceof RecipeChoice.ExactChoice){
            rec = new ShapedRecipe(NamespacedKey.minecraft("bmc_shaped_"+name.toLowerCase()+"_"+key),((RecipeChoice.ExactChoice)resultItemFromScheme.getValue()).getItemStack());
        } else if (resultItemFromScheme.getValue() instanceof RecipeChoice.MaterialChoice){
            rec = new ShapedRecipe(NamespacedKey.minecraft("bmc_shaped_"+name.toLowerCase()+"_"+key),((RecipeChoice.MaterialChoice)resultItemFromScheme.getValue()).getItemStack());
        } else {
            rec = new ShapedRecipe(NamespacedKey.minecraft("bmc_shaped_"+name.toLowerCase()+"_"+key),((RecipeChoice.ExactChoice)resultItemFromScheme.getValue()).getItemStack());
        }
        assert shape != null;
        for (int i = 0; i < shape.toArray().length; i++){
            String[] itemSchemes = shape.get(i).split("\\|");
            for (String itemScheme: itemSchemes) {
                Map.Entry<Character, RecipeChoice> itemFromScheme = BmcCraftSubContainer.convertSchemeToRecipeChoice(itemScheme, (char)(97+i));

                if(itemFromScheme == null){
                    BetterMinecraftCore.getInstance().getLogger().warning("§cskipping shaped recipe §3"+name+" "+key+"&c, some ingredients are invalid or not registered");
                    return;
                }
                shapeArr[i] += itemFromScheme.getKey();
                if(itemFromScheme.getValue() != null)
                    ingredients.put(itemFromScheme.getKey(), itemFromScheme.getValue());
            }
        }
        rec.shape(shapeArr[0],shapeArr[1],shapeArr[2]);
        for (Map.Entry<Character, RecipeChoice> ingredient: ingredients.entrySet()) {
            rec.setIngredient(ingredient.getKey(),ingredient.getValue());
        }
        BetterMinecraftCore.getInstance().getServer().addRecipe(rec);
    }

    public BmcShapedCraft cloneForConfig() {
        BmcShapedCraft sc = new BmcShapedCraft();
        sc.setKey(null);
        sc.setName(null);
        sc.setShape(shape);
        sc.setResult(result);
        return sc;
    }

    public void registerSelfInConfig(){
        Map<String, BmcCraftSubContainer> bmcCraftSubContainerMap = BmcCraftContainer.getInstance().getRecipes();
        if(bmcCraftSubContainerMap == null){
            bmcCraftSubContainerMap = new HashMap<>();
        }
        BmcCraftSubContainer bmcCraftSubContainer = bmcCraftSubContainerMap.get(name);
        if(bmcCraftSubContainer == null){
            bmcCraftSubContainer = new BmcCraftSubContainer();
        }
        Map<String, BmcShapedCraft> bmcCraftMap = bmcCraftSubContainer.getShaped();
        if(bmcCraftMap == null){
            bmcCraftMap = new HashMap<>();
        }
        BmcShapedCraft bmcCraft = bmcCraftMap.get(key);
        if(bmcCraft == null){
            bmcCraft = new BmcShapedCraft();
        }
        bmcCraft.setKey(key);
        bmcCraft.setName(name);
        bmcCraft.setShape(shape);
        bmcCraft.setResult(result);
        bmcCraftMap.put(key,bmcCraft);
        bmcCraftSubContainer.setShaped(bmcCraftMap);
        bmcCraftSubContainerMap.put(name,bmcCraftSubContainer);
        BmcCraftContainer.getInstance().setRecipes(bmcCraftSubContainerMap);
        BmcCraftContainer.save();
        BmcCraftContainer.registerCrafts();
    }
}

File: src/main/kotlin/me/rsman/BetterMinecraftCore/configs/models/BmcShapedCraft.kt
package me.rsman.BetterMinecraftCore.configs.models

import me.rsman.BetterMinecraftCore.BetterMinecraftCore.Companion.instance
import java.util.HashMap
import org.bukkit.inventory.RecipeChoice
import org.bukkit.inventory.ShapedRecipe
import org.bukkit.inventory.RecipeChoice.ExactChoice
import org.bukkit.NamespacedKey
import org.bukkit.inventory.RecipeChoice.MaterialChoice
import me.rsman.BetterMinecraftCore.configs.containers.BmcCraftContainer

class BmcShapedCraft {
    var shape: List<String>? = null
    var result: String? = null
    var name: String? = null
    var key: String? = null
    fun registerCraft() {
        val shapeArr = arrayOf("", "", "")
        val ingredients: MutableMap<Char, RecipeChoice?> = HashMap()
        assert(result != null)
        val resultItemFromScheme: Map.Entry<Char, RecipeChoice?>? = BmcCraftSubContainer.convertSchemeToRecipeChoice(result, 'X')
        if (resultItemFromScheme?.value == null) {
            instance.logger.warning("§cskipping shaped recipe §3$name $key§c, result is invalid or not registered")
            return
        }
        val rec: ShapedRecipe = when(resultItemFromScheme.value){
            is MaterialChoice -> ShapedRecipe(NamespacedKey.minecraft("bmc_shaped_" + name!!.lowercase() + "_" + key), (resultItemFromScheme.value as MaterialChoice?)!!.itemStack)
            is ExactChoice -> ShapedRecipe(NamespacedKey.minecraft("bmc_shaped_" + name!!.lowercase() + "_" + key), (resultItemFromScheme.value as ExactChoice?)!!.itemStack)
            else -> ShapedRecipe(NamespacedKey.minecraft("bmc_shaped_" + name!!.lowercase() + "_" + key), (resultItemFromScheme.value as ExactChoice?)!!.itemStack)
        }
        assert(shape != null)
        for (i in shape!!.toTypedArray().indices) {
            val itemSchemes = shape!![i].split("\\|".toRegex()).toTypedArray()
            for (itemScheme in itemSchemes) {
                val itemFromScheme: Map.Entry<Char, RecipeChoice?>? = BmcCraftSubContainer.convertSchemeToRecipeChoice(itemScheme, (97 + i).toChar())
                if (itemFromScheme == null) {
                    instance.logger.warning("§cskipping shaped recipe §3$name $key&c, some ingredients are invalid or not registered")
                    return
                }
                shapeArr[i] = shapeArr[i] + itemFromScheme.key
                if (itemFromScheme.value != null) ingredients[itemFromScheme.key] = itemFromScheme.value
            }
        }
        rec.shape(shapeArr[0], shapeArr[1], shapeArr[2])
        for ((key1, value) in ingredients) {
            rec.setIngredient(key1, value!!)
        }
        instance.server.addRecipe(rec)
    }

    fun cloneForConfig(): BmcShapedCraft {
        val sc = BmcShapedCraft()
        sc.key = null
        sc.name = null
        sc.shape = shape
        sc.result = result
        return sc
    }

    fun registerSelfInConfig() {
        var bmcCraftSubContainerMap: MutableMap<String?, BmcCraftSubContainer?>? = BmcCraftContainer.instance?.recipes
        if (bmcCraftSubContainerMap == null) {
            bmcCraftSubContainerMap = HashMap()
        }
        var bmcCraftSubContainer = bmcCraftSubContainerMap[name]
        if (bmcCraftSubContainer == null) {
            bmcCraftSubContainer = BmcCraftSubContainer()
        }
        var bmcCraftMap = bmcCraftSubContainer.shaped
        if (bmcCraftMap == null) {
            bmcCraftMap = HashMap()
        }
        var bmcCraft = bmcCraftMap[key]
        if (bmcCraft == null) {
            bmcCraft = BmcShapedCraft()
        }
        bmcCraft.key = key
        bmcCraft.name = name
        bmcCraft.shape = shape
        bmcCraft.result = result
        bmcCraftMap[key] = bmcCraft
        bmcCraftSubContainer.shaped = bmcCraftMap
        bmcCraftSubContainerMap[name] = bmcCraftSubContainer
        BmcCraftContainer.instance?.recipes = bmcCraftSubContainerMap
        BmcCraftContainer.save()
        BmcCraftContainer.registerCrafts()
    }
}